### DISCUSSION TITLE
Does setting BlendFunction.SKIP avoid calculating a pass, or just skip blending?

### DISCUSSION THREAD

**Discussion by looeee** [pmndrs/postprocessing#315] [Q&A] [answered]
If I create a pass and set `blendFunction = BlendFunction.SKIP`, does this skip the entire pass, or does it still calculate the pass but just skip blending it into the final result (i.e. wasted calculation)? 

If it does skip the entire pass, then if I've already created a postprocessing sequence, for example:

* ToneMapping
* Bloom
* SSAO
* Outline

If I see performance regression and set Bloom and SSAO to BlendFunction.SKIP, will that immediately skip the calculation of those passes? 
Or do I have to recreate the Bloom and SSAO with a new `blendFunction` setting?

**Comment by vanruesc** [comment, marked as answer]
Setting the blend function of an effect's [blendMode](https://vanruesc.github.io/postprocessing/public/docs/class/src/effects/Effect.js~Effect.html#instance-member-blendMode) via [setBlendFunction()](https://vanruesc.github.io/postprocessing/public/docs/class/src/effects/blending/BlendMode.js~BlendMode.html#instance-method-setBlendFunction) will always trigger a recompilation of the associated effect material. When the blend function is set to `BlendFunction.SKIP`, the effect's main fragment shader won't be integrated into the final shader at all, but the `update()` method of the effect will still be called. This is because the effect may render to offscreen render targets (like the bloom overlay) which could be used by other effects further down the pipeline.

> If I see performance regression and set Bloom and SSAO to BlendFunction.SKIP, will that immediately skip the calculation of those passes?

Not entirely. Bloom and SSAO won't show up in the effect shader, but they will still calculate their offscreen textures.

If you want to completely enable/disable effects on the fly, you should prepare multiple instances of `EffectPass` with the desired effect combinations ahead of time and enable/disable these passes as needed.

Skipping blending is mainly intended for more complex effect compositions that require deferred blending.

**Reply by looeee** [reply]
Got it, thank you. It seems like the passes that I'm most likely to want to disable (e.g. SSAO) probably all have offscreen textures. 

I did some testing with SSAO and it seems like there's minimal performance gain by setting it to SKIP - there's a big difference between removing the pass entirely and SKIP and no noticeable difference between MULTIPLY and SKIP. 

> Skipping blending is mainly intended for more complex effect compositions that require deferred blending.

Yeah, that makes sense - you mean like in the [Color Grading demo](https://vanruesc.github.io/postprocessing/public/demo/#color-grading) where you can enable/disable multiple effects?

**Reply by vanruesc** [reply]
> you mean like in the Color Grading demo where you can enable/disable multiple effects?

That demo is a special case because none of the color grading effects produce any offscreen textures, so skipping them effectively disables them. The small freezes caused by shader recompilations are acceptable in the context of this demo because they are actively triggered by the user.

I can't think of a good example where you'd want to skip blending. Maybe `SKIP` isn't even needed anymore since we now have selective bloom, but it was useful for some bloom setups in the past.

**Reply by looeee** [reply]
> Maybe SKIP isn't even needed anymore since we now have selective bloom, but it was useful for some bloom setups in the past.

Either that or clearly state in the docs/types that SKIP doesn't eliminate computation for passes with offscreen textures. Currently, it does look like that's what it would do - and examples like the Color Grading give that impression as you set a pass to SKIP to disable it. The [R3F examples](http://docs.pmnd.rs/react-three-fiber/examples) - [specifically this one](https://codesandbox.io/s/performance-scaling-pz0q6) - also use it this way. 

It was only once I tried actually implementing performance scaling that I got suspicious that it could work this way 😅

**Reply by vanruesc** [reply]
I've added a note to the [docs](https://vanruesc.github.io/postprocessing/public/docs/class/src/effects/Effect.js~Effect.html#instance-method-dispose) and updated the related [Wiki page](https://github.com/vanruesc/postprocessing/wiki/Enabling-and-Disabling-Effects). Not sure if I should change the color grading demo since those effects can actually be fully disabled this way.

**Reply by looeee** [reply]
Cool, the Wiki page is very clear. I wasn't suggesting changing the color grading demo btw, I think it's fine now that the docs are clearer.
